### DISCUSSION TITLE
How to install MicroPython to different flash size? (ESP_ERR_FLASH_NOT_INITIALISED)

### DISCUSSION THREAD

**Discussion by dusan-zervan** [micropython/micropython#10156] [Using MicroPython] [answered]
Hello, I am using MicroPython on many different boards, but now I have problem with the new one.. It is _LilyGO T-QT Pro_ with **ESP32-S3 and 4 MiB flash size**. It has also 2 MiB of SPIRAM, so I have used [Generic ESP32-S3 (SPIRAM) image](https://micropython.org/download/GENERIC_S3_SPIRAM/). It is working, but I can't save any file to device. It says an error after boot: `OSError: (-24579, 'ESP_ERR_FLASH_NOT_INITIALISED')`

It seems like the image is designed for 8 MiB flash size and when I have flashed it, device reports back incorrect size - `esp.flash_size()` reports 8388608 bytes. I am using esptool to flash, so I tried to use parameter `--flash_size 4MB` - no success, it is not accepted, because of this warning:
> Warning: Image file at 0x0 is protected with a hash checksum, so not changing the flash size setting.

Is there a way to solve that without compilation? (I have never compiled such project.)

**Comment by glenn20** [comment, marked as answer]
I believe you will need to recompile (unless someone else corrects me - or there are some precompiled 4MiB S3 images around):

You can set the flashsize in `ports/esp32/boards/GENERIC_S3/sdkconfig.board`  and select the default partition table for a 4MB flash:

```
...
CONFIG_ESPTOOLPY_FLASHSIZE_4MB=y
CONFIG_ESPTOOLPY_FLASHSIZE_8MB=
CONFIG_ESPTOOLPY_FLASHSIZE_16MB=
CONFIG_PARTITION_TABLE_CUSTOM=y
CONFIG_PARTITION_TABLE_CUSTOM_FILENAME="partitions.csv"
```

Compiling ain't so hard with docker (thanks to @mattytrentini):
```bash
git clone git@github.com:micropython/micropython.git
cd micropython
# Edit ports/esp32/boards/GENERIC_S3/sdkconfig.board as described above
docker run -ti --rm -v $(pwd):/micropython -w /micropython espressif/idf:release-v4.4 bash -c "make -C mpy-cross && make -C ports/esp32 submodules all BOARD=GENERIC_S3 -j4"
```

**Reply by mattytrentini** [reply]
(Apologies @glenn20, didn't mean to basically duplicate your response; it just hadn't appeared by the time I finally wrote my response!)

**Reply by unixb0y** [reply]
Worked for me, but not with `partitions.csv`, instead I used https://github.com/raibisch/ESP32-S3-WROOM-4MB/blob/master/4mb_2ota_partition.csv .

To be exact, I'm using the ESP32-S3-WROOM-1-N4 module.

**Reply by edcline** [reply]
I am having the same issue, just trying to do a basic prototype board bring-up with S3-WROOM-1U-N4, don't need a production version of mpy.  Would you be able to post your binary for a quick test?

**Reply by unixb0y** [reply]
@edcline yes, I uploaded it here: https://github.com/unixb0y/Micropython-ESP32-S3-WROOM-1-N4

**Reply by edcline** [reply]
Excellent, thank you...

**Comment by mattytrentini** [comment]
You're right; I had the same issue. 

It's only a straightforward change but it does necessitate compilation. In [GENERIC_S3_SPIRAM/sdkconfig.board](https://github.com/micropython/micropython/blob/master/ports/esp32/boards/GENERIC_S3_SPIRAM/sdkconfig.board) you need to make the following changes:

```ini
CONFIG_FLASHMODE_QIO=y
CONFIG_ESPTOOLPY_FLASHFREQ_80M=y
CONFIG_ESPTOOLPY_FLASHSIZE_DETECT=y
CONFIG_ESPTOOLPY_AFTER_NORESET=y

CONFIG_SPIRAM_MEMTEST=

CONFIG_ESPTOOLPY_FLASHSIZE_4MB=y # This was unset
CONFIG_ESPTOOLPY_FLASHSIZE_8MB= # This was set
CONFIG_ESPTOOLPY_FLASHSIZE_16MB=
CONFIG_PARTITION_TABLE_CUSTOM=y
CONFIG_PARTITION_TABLE_CUSTOM_FILENAME="partitions.csv" # This pointed to 'partitions-8MiB.csv'
```

I think we should follow-up with two options: a) Add a board definition for the [T-QT Pro](https://www.lilygo.cc/products/t-qt-pro) with this configuration. b) Add variants to allow multiple builds of the GENERIC_S3_SPIRAM for the common flash sizes...

**Reply by edcline** [reply]
Was this ever completed....is there a build available for the 4MB flash modules?

**Reply by FriendlyNGeeks** [reply]
> Was this ever completed....is there a build available for the 4MB flash modules?

What are you trying to flash, brand; model?

**Reply by edcline** [reply]
I've got an ESP32-S3-WROOM-1U-N4R2

**Comment by dusan-zervan** [comment]
Excellent, thank you both! I had to use `git clone https://github.com/micropython/micropython.git` instead of `git clone git@github.com:micropython/micropython.git` - I don't know why. But it is working.

However, I have some questions (maybe not very clever, I have never used that):
- I've got 3 files: _bootloader.bin_ to 0x0, _partition-table.bin_ to 0x8000 and _micropython.bin_ to 0x10000. Is there a way to modify some git parameters to combine those files to just one file?
- I've got a "nightly build" 724. Is there a way to get official / stable build?
- Why is it needed to use IDF 4.4? I don't even understand what is it (never mind), but I've tried idf:latest and it failed to compile.
- Perhaps I could also integrate some modules that I need, maybe even my own boot.py file directly to image. Could you direct me to some instructions / tutorial how to do that?

**Reply by dusan-zervan** [reply]
For my last question I have found information:
- https://docs.micropython.org/en/latest/reference/manifest.html
- https://github.com/micropython/micropython/wiki/Building-Micropython-Binaries

It is good starting point, I will study that later.

**Comment by glenn20** [comment]
> Excellent, thank you both! I had to use `git clone https://github.com/micropython/micropython.git` instead of `git clone git@github.com:micropython/micropython.git` - I don't know why. But it is working.

Whoops, my version is for ssh auth to GitHub, which is what I use, but only useful if you have setup SSH access.

> However, I have some questions (maybe not very clever, I have never used that):
> - I've got 3 files: _bootloader.bin_ to 0x0, _partition-table.bin_ to 0x8000 and _micropython.bin_ to 0x10000. Is there a way to modify some git parameters to combine those files to just one file?
> - I've got a "nightly build" 724. Is there a way to get official / stable build?

Yes, from within your micropython folder execute `git checkout v1.19` to get the latest micropython release tagged version. Then run the build docker command again. You should probably run the "make clean" docker command first.

> - Why is it needed to use IDF 4.4? I don't even understand what is it (never mind), but I've tried idf:latest and it failed to compile.

This is the idf version that is used to build the production  esp32s2/3 images. The 5.x idf is only recently released.

The IDF is the integrated development framework provided by espressif, the manufacturer of the esp32 devices. The Micropython esp32 port is built on top of that.

> - Perhaps I could also integrate some modules that I need, maybe even my own boot.py file directly to image. Could you direct me to some instructions / tutorial how to do that?

Looks like you found the answer to that one with the manifest file.

**Comment by steeley** [comment]
I have an ESP32-s3 4R8 and tried everything here and while I can use the flashed firmware I have 0 storage so can't save anything.
Anyone got this config working?
module is this
[[(https://www.digikey.co.uk/en/products/detail/espressif-systems/ESP32-S3-WROOM-1U-N4R8/16162634](https://github.com/orgs/micropython/discussions/url)](url)

The module works perfectly in Arduino  so my hardware setup is just fine.

**Comment by bixb922** [comment]
I don't have that specific module, but this utility program; https://github.com/glenn20/mp-image-tool-esp32 allows you to set the flash size without needing to recompile Micropython. It worked well for me for an ESP32-S3 N8R8 and N16R8 with SPIRAM starting with the standard image for octal SPIRAM from https://micropython.org/download/ESP32_GENERIC_S3/

If not, you will have to install Esp-idf and recompile Micropython, see https://github.com/micropython/micropython/blob/master/ports/esp32/README.md

**Reply by steeley** [reply]
Many thanks- this is a useful tool. seems to work, but does produce some errors.
Did you have similar errors?
`
>>> ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0x1 (POWERON),boot:0x8 (SPI_FAST_FLASH_BOOT)
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fce3810,len:0xf8c
load:0x403c9700,len:0xb3c
load:0x403cc700,len:0x2dd4
SHA-256 comparison failed:
Calculated: 2aa7a6e666276c5e883d1857335d26b255b8084d17eb005758125e5a09cccfe0
Expected: 8a44d662a8bf8b449e4fed0ba7de4e2f4b89af831d9e77d09a76c993bbbf1a37
Attempting to boot anyway...
entry 0x403c989c
E (306) quad_psram: PSRAM ID read error: 0x00ffffff, PSRAM chip not found or not supported, or wrong PSRAM line mode
E (306) esp_psram: PSRAM enabled but initialization failed. Bailing out.
MicroPython v1.22.1 on 2024-01-05; Generic ESP32S3 module with ESP32S3
Type "help()" for more information.`

**Reply by bixb922** [reply]
Are you sure you downloaded the firmware from the section of the download page marked as "Firmware (Support for Octal-SPIRAM)"? The name of the Micropython version in the log seems to indicate that this is the version without SPIRAM support?

**Reply by steeley** [reply]
Ahh yes - accidentally converted the wrong image. the SPIRAM version seems ok now, but I still get the SHA checksum error  - not sure if thats a problem?

**Reply by bixb922** [reply]
The SHA comparison error could means that the Micropython image arrived corrupted at the ESP32-S3. 

Could you please post the command you are using to flash? Try using esptool with -z, the compression adds also a certain layer of transmission checking

Have you tried erase flash before writing?

Do you get the REPL prompt?

Are you using the serial port or the USB native port of the ESP32S3? With the native USB port I get less transmission errors than with the serial port

**Reply by robert-hh** [reply]
@steeley I never have seen 'real' SHA1 errors after uploading a ESP32 image. So I suppose that the tool you use to adapt for a 4MB flash did not recreate the SHA1 checksum of the firmware file. Then the SHA1 checksum of the file does not match.

**Reply by steeley** [reply]
thanks for the suggestions - REPL did work OK. I'm using the serial via a known good usb-UART serial adapter cable.
I always erase flash before writing.
Yes- could use the built-in USB on the S3  -in fact I'm going to add a connector for this.

I'll try esptool directly with -z. I was using Thonny to upload which uses esptool but without the -z option I think.

**Comment by a-sapuppo** [comment]
we use ESP32-S3-WROOM-1U-N16R8 for micropython
